index.bs

```diff
@@ -1772,6 +1963,278 @@ contexts that have active documents; navigation can also cause contexts to
 become inaccessible but not yet get discarded because bfcache.
 </div>
 
+#### The browsingContext.navigationStarted Event #### {#event-browsingContext-navigationStarted}
+
+<dl>
+   <dt>Event Type</dt>
+   <dd>
+      <pre class="cddl local-cddl">
+        BrowsingContextNavigationStartedEvent = {
+         method: "browsingContext.navigationStarted",
+         params: BrowsingContextNavigationInfo
+       }
+      </pre>
+   </dd>
+</dl>
+The [=remote end event trigger=] is:
+
+<div algorithm="remote end event trigger for browsingContext.navigationStarted">
+
+Define the following [=navigation started steps=] with |context| and |navigation status|:
+
+ 1. If the [=current session=] is null, return.
+
+ 1. Let |context id| be the [=browsing context id=] for |context|.
+
+ 1. Let |navigation id| be |navigation status|'s navigation id.
+
+ 1. Let |url| be |navigation status|'s url.
+
+ 1. Let |params| be a [=map=] matching the
+    <code>BrowsingContextNavigationInfo</code> production, with the
+    <code>context</code> field set to |context id|, the <code>navigation</code>
+    field set to |navigation id| and the <code>url</code> field set to the
+    result of the [=URL serializer=] given |url|.
+
+ 1. Let |body| be a [=map=] matching the
+     <code>BrowsingContextNavigationStarted</code> production, with the
+     <code>params</code> field set to |params|.
+
+ 1. Let |related browsing contexts| be a set containing |context|.
+
+ 1. [=Emit an event=] with |body| and |related browsing contexts|.
```

Comment:
> I lean towards making the command response go first. Another option would be to have the command take an ID given by the client, but that seems like too big a hammer for the problem.

I'm also leaning towards making the command response go first when `wait` is not used, but I suspect this may not be a big issue. Clients that want to implement custom wait logic on top of BiDi are likely going to use the nav ID from the command response and can ignore the navigationStarted.

Reply:
Let's try to ensure the command response goes first then to not have to revisit this issue in the future.

Taking another look at https://pr-preview.s3.amazonaws.com/w3c/webdriver-bidi/pull/93.html#command-browsingContext-navigate it looks like step 18 ('If wait condition is "none"') already does the right thing.

@jgraham just to confirm, you intended to make this change and you too think it works now?

Reply:
I think in the navigation started case we still send the event first and then return since the way the spec is written everything is synchronous. In the fragment case we suppress the event and re-emit it. I'll update to do the same here, but I think we need to replace this with an explicit task based system. The current task spec, however, assumes that there is an ecmascript agent associated with the event loop; it's somewhat unclear if we can assume that for stuff that's conceptually happening in the parent/UI process (in Firefox, of course, there is exactly such an agent because we run js for the UI. But I don't know if that's something we can assume in general).

Reply:
I filed https://github.com/w3c/webdriver-bidi/issues/113 to work out the details of the formalism here.